### PR TITLE
Create analytics on the route given in http request

### PR DESCRIPTION
**Pull request by jiminbyun**
Given the array of waypoints:
- Generate rectangle bound that encapsulates all of the points
- Range the quadtree on that rectangular bound
- From the queried reports, get all reports near line
- Create analytics from reports near line

While deciding whether reports are near the line or not, plane geometry was used treating longitude and latitude as cartesian product because the routes will mostly be local area less than an hour drive and encorporating sphere geometry uses lots of trig functions that will increase runtime a lot as checking whether point is near the line is called multiple times.

### CODE REVIEW

File: web-app/src/main/java/com/google/sps/data/Distance.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      if (longitude < minLongitude) {
+        minLongitude = longitude;
+      }
+    }
+
+    return new Rectangle(new Coordinates(maxLatitude, minLongitude), new Coordinates(minLatitude, maxLongitude));
+  }
+
+  private static double square(double num) {
+    return (num*num);
```
Comment (AshleyLTW) [inline web-app/src/main/java/com/google/sps/data/Distance.java]: nit: white space on either side of *

File: web-app/src/main/java/com/google/sps/data/Distance.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  private static double distanceSquared(Coordinates start, Coordinates end) {
+    return square(start.getLng()-end.getLng()) + square(start.getLat()-end.getLat());
```
Comment (AshleyLTW) [inline web-app/src/main/java/com/google/sps/data/Distance.java]: nit: white space on either side of -

File: web-app/src/main/java/com/google/sps/data/Distance.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  public static double distanceSquaredFromSegment(Coordinates start, Coordinates end, Coordinates point) {
+    double segmentDistanceSquared = distanceSquared(start, end);
+
+    if (segmentDistanceSquared < 1e-10) {
+      return distanceSquared(start, point);
+    }
+
+    double projectionScale = 
+      ((point.getLng() - start.getLng()) * (end.getLng()-start.getLng()) + (point.getLat() - start.getLat()) * (end.getLat()-start.getLat()))
```
Comment (AshleyLTW) [inline web-app/src/main/java/com/google/sps/data/Distance.java]: nit: whitespace on either side of -

**Review by AshleyLTW — APPROVED** [review]

File: web-app/src/main/java/com/google/sps/servlets/AnalyticsServlet.java [hunk 4]
```diff
@@ -56,6 +59,21 @@ public void init() {
     }
   }
 
+  private boolean isReportNearLine(PoliceReport report, Coordinates[] waypoints) {
+    Coordinates reportLocation = new Coordinates(report.getLat(), report.getLng());
+    int index = 0;
+    while (index < waypoints.length - 1) {
```
Comment (darrnshn) [inline web-app/src/main/java/com/google/sps/servlets/AnalyticsServlet.java]: nit: for loop?

**Review by darrnshn — APPROVED** [review]

**Pull request merged**